File: ParticleNet/python/DataFormat.py
from ROOT import TLorentzVector
from ROOT import TMath

# Base class for all objects
class Particle(TLorentzVector):
    def __init__(self, pt, eta, phi, mass):
        TLorentzVector.__init__(self)
        self.SetPtEtaPhiM(pt, eta, phi, mass)
        self.charge = 0
        self.btagScore = 0.
        self.isMuon = False
        self.isElectron = False
        self.isJet = False

    def Charge(self):
        return self.charge

    def BtagScore(self):
        return self.btagScore

    def MT(self, part):
        dPhi = self.DeltaPhi(part)
        return TMath.Sqrt(2*self.Pt()*part.Pt()*(1.-TMath.Cos(dPhi)))

    def IsMuon(self):
        return self.isMuon

    def IsElectron(self):
        return self.isElectron

    def IsJet(self):
        return self.isJet

# Base class for electron / muon
class Lepton(Particle):
    def __init__(self, pt, eta, phi, mass):
        Particle.__init__(self, pt, eta, phi, mass)

    def SetCharge(self, charge):
        self.charge = charge


class Muon(Lepton):
    def __init__(self, pt, eta, phi, mass):
        Lepton.__init__(self, pt, eta, phi, mass)
        self.isMuon = True


class Electron(Lepton):
    def __init__(self, pt, eta, phi, mass):
        Lepton.__init__(self, pt, eta, phi, mass)
        self.isElectron = True


class Jet(Particle):
    def __init__(self, pt, eta, phi, mass):
        Particle.__init__(self, pt, eta, phi, mass)
        self.isJet = True
        self.isBtagged = False

    def SetCharge(self, charge):
        self.charge = charge

    def SetBtagScore(self, btagScore):
        self.btagScore = btagScore

    def SetBtagging(self, isBtagged):
        self.isBtagged = isBtagged

    def IsBtagged(self):
        return self.isBtagged

def getMuons(evt):
    muons = []
    muons_zip = zip(evt.MuonPtColl,
                    evt.MuonEtaColl,
                    evt.MuonPhiColl,
                    evt.MuonMassColl,
                    evt.MuonChargeColl)
    for pt, eta, phi, mass, charge in muons_zip:
        thisMuon = Muon(pt, eta, phi, mass)
        thisMuon.SetCharge(charge)
        muons.append(thisMuon)
    return muons

def getElectrons(evt):
    electrons = []
    electrons_zip = zip(evt.ElectronPtColl,
                        evt.ElectronEtaColl,
                        evt.ElectronPhiColl,
                        evt.ElectronMassColl,
                        evt.ElectronChargeColl)
    for pt, eta, phi, mass, charge in electrons_zip:
        thisElectron = Electron(pt, eta, phi, mass)
        thisElectron.SetCharge(charge)
        electrons.append(thisElectron)
    return electrons

def getJets(evt):
    jets = []
    jets_zip = zip(evt.JetPtColl,
                   evt.JetEtaColl,
                   evt.JetPhiColl,
                   evt.JetMassColl,
                   evt.JetChargeColl,
                   evt.JetBtagScoreColl,
                   evt.JetIsBtaggedColl)
    for pt, eta, phi, mass, charge, btagScore, isBtagged in jets_zip:
        thisJet = Jet(pt, eta, phi, mass)
        thisJet.SetCharge(charge)
        thisJet.SetBtagScore(btagScore)
        thisJet.SetBtagging(isBtagged)
        jets.append(thisJet)

    bjets = list(filter(lambda jet: jet.IsBtagged(), jets))
    return jets, bjets

File: ParticleNet/src/DataFormat.cc
#include "DataFormat.h"

// Particle class implementations
Particle::Particle() : TLorentzVector(), charge(0), btagScore(0), isMuon(false), isElectron(false), isJet(false) {}

Particle::Particle(Double_t pt, Double_t eta, Double_t phi, Double_t mass) : TLorentzVector() {
    SetPtEtaPhiM(pt, eta, phi, mass);
    charge = 0;
    btagScore = 0;
    isMuon = false;
    isElectron = false;
    isJet = false;
}

Float_t Particle::Charge() const { return charge; }
Float_t Particle::BtagScore() const { return btagScore; }
Bool_t Particle::IsMuon() const { return isMuon; }
Bool_t Particle::IsElectron() const { return isElectron; }
Bool_t Particle::IsJet() const { return isJet; }

Double_t Particle::MT(const Particle& part) const {
    Double_t dPhi = DeltaPhi(part);
    return TMath::Sqrt(2 * Pt() * part.Pt() * (1. - TMath::Cos(dPhi)));
}

void Particle::SetCharge(Float_t c) { charge = c; }
void Particle::SetBtagScore(Float_t score) { btagScore = score; }

// Lepton class implementations
Lepton::Lepton(Double_t pt, Double_t eta, Double_t phi, Double_t mass) : Particle(pt, eta, phi, mass) {}

// Muon class implementations
Muon::Muon(Double_t pt, Double_t eta, Double_t phi, Double_t mass) : Lepton(pt, eta, phi, mass) {
    isMuon = true;
}

// Electron class implementations
Electron::Electron(Double_t pt, Double_t eta, Double_t phi, Double_t mass) : Lepton(pt, eta, phi, mass) {
    isElectron = true;
}

// Jet class implementations
Jet::Jet(Double_t pt, Double_t eta, Double_t phi, Double_t mass) : Particle(pt, eta, phi, mass) {
    isJet = true;
    isBtagged = false;
}

void Jet::SetBtagging(Bool_t tagged) { isBtagged = tagged; }
Bool_t Jet::IsBtagged() const { return isBtagged; }

// EventData implementations
void EventData::SetBranches(TTree* tree) {
    // Muon branches
    tree->SetBranchAddress("MuonPtColl", &MuonPtColl);
    tree->SetBranchAddress("MuonEtaColl", &MuonEtaColl);
    tree->SetBranchAddress("MuonPhiColl", &MuonPhiColl);
    tree->SetBranchAddress("MuonMassColl", &MuonMassColl);
    tree->SetBranchAddress("MuonChargeColl", &MuonChargeColl);
    
    // Electron branches
    tree->SetBranchAddress("ElectronPtColl", &ElectronPtColl);
    tree->SetBranchAddress("ElectronEtaColl", &ElectronEtaColl);
    tree->SetBranchAddress("ElectronPhiColl", &ElectronPhiColl);
    tree->SetBranchAddress("ElectronMassColl", &ElectronMassColl);
    tree->SetBranchAddress("ElectronChargeColl", &ElectronChargeColl);
    
    // Jet branches
    tree->SetBranchAddress("JetPtColl", &JetPtColl);
    tree->SetBranchAddress("JetEtaColl", &JetEtaColl);
    tree->SetBranchAddress("JetPhiColl", &JetPhiColl);
    tree->SetBranchAddress("JetMassColl", &JetMassColl);
    tree->SetBranchAddress("JetChargeColl", &JetChargeColl);
    tree->SetBranchAddress("JetBtagScoreColl", &JetBtagScoreColl);
    tree->SetBranchAddress("JetIsBtaggedColl", &JetIsBtaggedColl);
    
    // Event variables
    tree->SetBranchAddress("METvPt", &METvPt);
    tree->SetBranchAddress("METvPhi", &METvPhi);
    tree->SetBranchAddress("nJets", &nJets);
}

// Helper functions to extract particles from event data
RVec<Muon> getMuons(const EventData& evt) {
    RVec<Muon> muons;
    if (!evt.MuonPtColl) return muons;
    
    size_t nMuons = evt.MuonPtColl->size();
    muons.reserve(nMuons);
    for (size_t i = 0; i < nMuons; ++i) {
        Muon mu((*evt.MuonPtColl)[i], 
                (*evt.MuonEtaColl)[i], 
                (*evt.MuonPhiColl)[i], 
                (*evt.MuonMassColl)[i]);
        mu.SetCharge((*evt.MuonChargeColl)[i]);
        muons.push_back(mu);
    }
    return muons;
}

RVec<Electron> getElectrons(const EventData& evt) {
    RVec<Electron> electrons;
    if (!evt.ElectronPtColl) return electrons;
    
    size_t nElectrons = evt.ElectronPtColl->size();
    electrons.reserve(nElectrons);
    for (size_t i = 0; i < nElectrons; ++i) {
        Electron el((*evt.ElectronPtColl)[i], 
                    (*evt.ElectronEtaColl)[i], 
                    (*evt.ElectronPhiColl)[i], 
                    (*evt.ElectronMassColl)[i]);
        el.SetCharge((*evt.ElectronChargeColl)[i]);
        electrons.push_back(el);
    }
    return electrons;
}

std::pair<RVec<Jet>, RVec<Jet>> getJets(const EventData& evt) {
    RVec<Jet> jets;
    RVec<Jet> bjets;
    if (!evt.JetPtColl) return std::make_pair(jets, bjets);
    
    size_t nJets = evt.JetPtColl->size();
    jets.reserve(nJets);
    for (size_t i = 0; i < nJets; ++i) {
        Jet jet((*evt.JetPtColl)[i], 
                (*evt.JetEtaColl)[i], 
                (*evt.JetPhiColl)[i], 
                (*evt.JetMassColl)[i]);
        jet.SetCharge((*evt.JetChargeColl)[i]);
        jet.SetBtagScore((*evt.JetBtagScoreColl)[i]);
        jet.SetBtagging((*evt.JetIsBtaggedColl)[i]);
        
        jets.push_back(jet);
        if (jet.IsBtagged()) {
            bjets.push_back(jet);
        }
    }
    return std::make_pair(jets, bjets);
}

// Convert particles to feature array for ML
RVec<RVec<float>> particlesToFeatures(const RVec<Muon>& muons,
                                      const RVec<Electron>& electrons,
                                      const RVec<Jet>& jets,
                                      const Particle& met) {
    RVec<RVec<float>> features;
    features.reserve(muons.size() + electrons.size() + jets.size() + 1); // +1 for MET
    
    // Add muons
    for (const auto& mu : muons) {
        features.push_back({
            static_cast<float>(mu.E()), 
            static_cast<float>(mu.Px()), 
            static_cast<float>(mu.Py()), 
            static_cast<float>(mu.Pz()),
            mu.Charge(), 
            mu.BtagScore(),
            1.0f, // IsMuon
            0.0f, // IsElectron
            0.0f  // IsJet
        });
    }
    
    // Add electrons
    for (const auto& el : electrons) {
        features.push_back({
            static_cast<float>(el.E()), 
            static_cast<float>(el.Px()), 
            static_cast<float>(el.Py()), 
            static_cast<float>(el.Pz()),
            el.Charge(), 
            el.BtagScore(),
            0.0f, // IsMuon
            1.0f, // IsElectron
            0.0f  // IsJet
        });
    }
    
    // Add jets
    for (const auto& jet : jets) {
        features.push_back({
            static_cast<float>(jet.E()), 
            static_cast<float>(jet.Px()), 
            static_cast<float>(jet.Py()), 
            static_cast<float>(jet.Pz()),
            jet.Charge(), 
            jet.BtagScore(),
            0.0f, // IsMuon
            0.0f, // IsElectron
            1.0f  // IsJet
        });
    }
    
    // Add MET
    features.push_back({
        static_cast<float>(met.E()), 
        static_cast<float>(met.Px()), 
        static_cast<float>(met.Py()), 
        static_cast<float>(met.Pz()),
        0.0f, // Charge
        0.0f, // BtagScore
        0.0f, // IsMuon
        0.0f, // IsElectron
        0.0f  // IsJet
    });
    
    return features;
}